### ISSUE TITLE
Sequence encoder decoder

### ISSUE THREAD

**Issue by basma-b** [datalogue/keras-attention#17]
I am using your decoder to implement my sequence encoder/decoder but actually i don't know how can I do to get the decoder output the same shape as my input. My input is (None, MAX_SEQ_LENGTH) these are my dialogue turns that I encode and the decode to get the next dialogue turn which has the same shape i.e (None, MAX_SEQ_LENGTH) although the decoder returns 3D dimension tensor because of return_sequences=True.

Can you help me please to this ?

**Comment by basma-b** [comment]
Here is a small code

``

    seq2seq = Sequential() # my turn shape=(None, MAX_SEQUENCE_LENGTH)
    seq2seq.add(Embedding(output_dim=args.emb_dim,
                            input_dim=MAX_NB_WORDS,
                            input_length=MAX_SEQUENCE_LENGTH,
                            weights=[embedding_matrix],
                            mask_zero=True,
                            trainable=True))
    seq2seq.add(LSTM(units=args.hidden_size, return_sequences=True))
    seq2seq.add(AttentionDecoder(args.hidden_size, args.emb_dim)) # the decoded shape=(None, MAX_SEQUENCE_LENGTH, args.emb_dim)
    seq2seq.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['acc'])

**Comment by zafarali** [comment]
Why is the shape (None,..)? Keras takes into account the batch size
internally!

**Comment by basma-b** [comment]
No sure no, this is what is actually been done inside. It's just to explain but you can ignore it ..

**Comment by zafarali** [comment]
Ok. Have you tried using the graph API instead of sequential?

**Comment by basma-b** [comment]
No, but can you give me more hints please ?

**Comment by zafarali** [comment]
A good intro is here: https://keras.io/getting-started/functional-api-guide/

I believe that the sequential API is just there for legacy purposes and
should not be used anymore.


This is an example of me using the graph API:
https://github.com/datalogue/keras-attention/blob/master/models/NMT.py

Does that help?

**Comment by basma-b** [comment]
Well I think the problem was not clear, but actually now I found a way to transform my decoder output to classes instead of probabilities using argmax. So now my output has the shape (MAX_SEQUENCE_LENGTH, ) the same as my inputs (I removed the batch size here).

Now I have another question: since my outputs are classes instead of probabilities which loss function could I use in this case.
example : 
output labels = [0, 0, 1, 12, 165, 3]
predicted_labels = [0, 0, 13, 166, 3]

**Comment by zafarali** [comment]
Since the output is from a categorical distribution you should use the
categorical cross entropy. You need the probabilities for this though. You
willl need to convert your predicted labels into one hot vectors
representing the class

**Comment by Ap1075** [comment]
For one hot encoding you could use the to_categorical function in keras, look it up in the documentation. To link your encoder to the attention decoder, you could use the RepeatVector layer to change the dimensions according to your requirements. Hope this helps.
